plugins/genesys/fu-genesys-hubhid-device.c

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright (C) 2023 Adam.Chen <Adam.Chen@genesyslogic.com.tw>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-genesys-common.h"
+#include "fu-genesys-hubhid-device.h"
+
+#define GENESYS_HUBHID_REPORT_ID	  0
+#define GENESYS_HUBHID_REPORT_BYTE_LENGTH 0x40
+#define GENESYS_HUBHID_REPORT_TIMEOUT	  100 /* ms */
+#define GENESYS_HUBHID_REPORT_FLAGS	  FU_HID_DEVICE_FLAG_ALLOW_TRUNC | FU_HID_DEVICE_FLAG_IS_FEATURE
+
+struct _FuGenesysHubhidDevice {
+	FuHidDevice parent_instance;
+	GByteArray *buf_report;
+};
+
+G_DEFINE_TYPE(FuGenesysHubhidDevice, fu_genesys_hubhid_device, FU_TYPE_HID_DEVICE)
+
+gboolean
+fu_genesys_hubhid_device_command_read(FuGenesysHubhidDevice *self,
+				      FuGenesysUsbSetup *setup,
+				      guint8 *data,
+				      FuProgress *progress,
+				      GError **error)
+{
+	FuHidDevice *hid_device = FU_HID_DEVICE(self);
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* set request data */
+	memset(self->buf_report->data, 0, self->buf_report->len);
+	if (!fu_memcpy_safe(self->buf_report->data,
+			    self->buf_report->len,
+			    0, /* dst */
+			    (const guint8 *)setup,
+			    sizeof(FuGenesysUsbSetup),
+			    0x0, /* src */
+			    sizeof(FuGenesysUsbSetup),
+			    error))
+		return FALSE;
+
+	/* send request report */
+	if (!fu_hid_device_set_report(hid_device,
+				      GENESYS_HUBHID_REPORT_ID,
+				      self->buf_report->data,
+				      self->buf_report->len,
+				      GENESYS_HUBHID_REPORT_TIMEOUT,
+				      GENESYS_HUBHID_REPORT_FLAGS,
+				      error))
+		return FALSE;
+
+	if (setup->length == 0) {
+		g_warning("read zero-length hid report");
+		return TRUE;
+	}
+
+	/* receive report */
+	chunks = fu_chunk_array_mutable_new(data, setup->length, 0, 0x0, self->buf_report->len);
+	if (progress != NULL) {
+		fu_progress_set_id(progress, G_STRLOC);
+		fu_progress_set_steps(progress, chunks->len);
+	}
```

Comment:
I think we should always use a FuProgress

Reply:
I try to modify to give each command with a FuProgress, and I have a question. Does it need a FuProgress if read only one byte?

Reply:
depends on how long that read takes :) If it's less than 1ms then don't bother.

Reply:
Still valid, no?

Reply:
No. I already checked all the R/W command, and add the `FuProgress` if it takes more than 1ms.